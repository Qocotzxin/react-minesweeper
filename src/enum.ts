export enum Piece {
  uncovered = "â–¡",
  bomb = "*",
}

export enum Message {
  lose = "You lose",
  ok = "OK",
  win = "You win",
}
