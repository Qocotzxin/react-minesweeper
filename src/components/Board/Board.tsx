import { Box, Stack, useMediaQuery } from "@mui/material";
import { FC, memo } from "react";
import { HiFlag } from "react-icons/hi";
import { useSelector } from "react-redux";
import { FixedSizeGrid } from "react-window";
import { RootState } from "../../app/store";
import { Message, Piece } from "../../enum";
import { DefaultTile, UncoveredTile } from "../../features";

interface BoardProps {
  boardMap: string[];
  difficulty: number;
  message: Message;
}

export const Board: FC<BoardProps> = ({ boardMap, difficulty, message }) => {
  const isDesktop = useMediaQuery("(min-width:680px)");
  const columns = (boardMap.length && boardMap[0].length) || 0;
  const availableFlags = useSelector(
    (state: RootState) => state.tile.availableFlags
  );

  const ItemRenderer = memo(({ style, rowIndex, columnIndex }: any) => {
    const content = boardMap[rowIndex][columnIndex];

    return (
      <Stack
        key={`tile-${rowIndex}-${columnIndex}`}
        direction="row"
        style={style}
      >
        {content === Piece.uncovered ? (
          <DefaultTile
            colIndex={columnIndex}
            rowIndex={rowIndex}
            disabled={message !== Message.ok}
          />
        ) : (
          <UncoveredTile key={columnIndex}>{content}</UncoveredTile>
        )}
      </Stack>
    );
  });

  return (
    <>
      <Stack direction="row" justifyContent="center" alignItems="center">
        <HiFlag
          size={18}
          color="rgb(201, 91, 118)"
          data-testid="DefaultTile-flag"
        />{" "}
        {availableFlags}
      </Stack>
      <Stack
        data-testid="Board"
        alignItems="center"
        my="1rem"
        width="max-content"
        maxWidth="100%"
        position="relative"
        // Max height: desktop - 30 rows, mobile - 15 rows
        maxHeight={isDesktop ? "480px" : "240px"}
        border={boardMap.length ? "4px solid whitesmoke" : "none"}
        borderRadius="4px"
        sx={{
          overflow: "auto",
        }}
      >
        {message !== Message.ok && (
          <Box
            width="100%"
            height="100%"
            bgcolor="rgba(0, 0, 0, 0.5)"
            position="absolute"
            top="0"
            left="0"
            zIndex="1"
            data-testid="Board-overlay"
          />
        )}
        <FixedSizeGrid
          columnCount={columns}
          rowCount={boardMap.length}
          columnWidth={24}
          rowHeight={24}
          height={difficulty === 1 ? 240 : 480}
          width={difficulty === 1 ? 240 : 480}
        >
          {ItemRenderer}
        </FixedSizeGrid>
      </Stack>
    </>
  );
};
