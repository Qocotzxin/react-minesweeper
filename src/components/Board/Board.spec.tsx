import { useMediaQuery } from "@mui/material";
import { render, screen } from "@testing-library/react";
import { Message, Piece } from "../../enum";
import { Board } from "./Board";
import tileReducer from "../../features/Tiles/tileSlice";
import { FC } from "react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";

const boardMockProps = {
  boardMap: [
    `${Piece.uncovered}${Piece.uncovered}${Piece.uncovered}${Piece.uncovered}`,
  ],
  difficulty: 1,
  message: Message.ok,
};

jest.mock("@mui/material", () => ({
  ...jest.requireActual("@mui/material"),
  useMediaQuery: jest.fn(() => false),
}));

afterAll(() => {
  jest.resetAllMocks();
  jest.resetModules();
});

const defaultState = {
  availableFlags: 10,
  flaggedTiles: [],
};

const Wrapper: FC = ({ children }) => {
  return (
    <Provider
      store={configureStore({
        reducer: { tile: tileReducer },
        preloadedState: { tile: defaultState },
      })}
    >
      {children}
    </Provider>
  );
};

describe("Board", () => {
  describe("Snapshot", () => {
    it("Should render correctly and match snapshot.", () => {
      const { container } = render(
        <Wrapper>
          <Board {...boardMockProps} />
        </Wrapper>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe("Behavior", () => {
    it("Should render a DefaultTile component for each string equal to â–¡.", () => {
      render(
        <Wrapper>
          <Board {...boardMockProps} />
        </Wrapper>
      );

      const defaultButtons = screen.getAllByTestId("DefaultTile");
      const UncoveredTiles = screen.queryAllByTestId("UncoveredTile");

      expect(defaultButtons.length).toEqual(4);
      expect(UncoveredTiles.length).toEqual(0);
    });

    it("Should render a UncoveredTile component for each string NOT equal to â–¡.", () => {
      render(
        <Wrapper>
          <Board boardMap={["1"]} difficulty={1} message={Message.ok} />
        </Wrapper>
      );

      const defaultButtons = screen.queryAllByTestId("DefaultTile");
      const UncoveredTiles = screen.getAllByTestId("UncoveredTile");

      expect(defaultButtons.length).toEqual(0);
      expect(UncoveredTiles.length).toEqual(1);
    });

    it("Should set the max height of the board to 15 rows in mobile resolutions.", () => {
      render(
        <Wrapper>
          <Board {...boardMockProps} />
        </Wrapper>
      );

      expect(screen.getByTestId("Board")).toHaveStyle({
        "max-height": "240px",
      });
    });

    it("Should set the max height of the board to 30 rows in desktop resolutions.", () => {
      (useMediaQuery as jest.Mock).mockImplementation(() => true);
      render(
        <Wrapper>
          <Board {...boardMockProps} />
        </Wrapper>
      );

      expect(screen.getByTestId("Board")).toHaveStyle({
        "max-height": "480px",
      });
    });

    it("Should have a border if the board map have elements.", () => {
      render(
        <Wrapper>
          <Board {...boardMockProps} />
        </Wrapper>
      );

      expect(screen.getByTestId("Board")).toHaveStyle({
        border: "4px solid whitesmoke",
      });
    });

    it("Should NOT have a border if the board map is empty.", () => {
      render(
        <Wrapper>
          <Board boardMap={[]} difficulty={1} message={Message.ok} />
        </Wrapper>
      );

      expect(screen.getByTestId("Board")).toHaveStyle({
        border: "none",
      });
    });

    it("Should display an overlay when the game is over.", () => {
      render(
        <Wrapper>
          <Board boardMap={[]} difficulty={1} message={Message.win} />
        </Wrapper>
      );

      expect(screen.getByTestId("Board-overlay")).toBeInTheDocument();
    });
  });
});
