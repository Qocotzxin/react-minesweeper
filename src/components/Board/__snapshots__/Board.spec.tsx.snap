// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board Snapshot Should render correctly and match snapshot. 1`] = `
<div>
  <div
    class="css-7owdng-MuiStack-root"
  >
    <svg
      color="rgb(201, 91, 118)"
      data-testid="DefaultTile-flag"
      fill="currentColor"
      height="18"
      stroke="currentColor"
      stroke-width="0"
      style="color: rgb(201, 91, 118);"
      viewBox="0 0 20 20"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M3 6a3 3 0 013-3h10a1 1 0 01.8 1.6L14.25 8l2.55 3.4A1 1 0 0116 13H6a1 1 0 00-1 1v3a1 1 0 11-2 0V6z"
        fill-rule="evenodd"
      />
    </svg>
     
    10
  </div>
  <div
    class="css-sdzd41-MuiStack-root"
    data-testid="Board"
  >
    <div
      style="position: relative; height: 240px; width: 240px; overflow: auto; will-change: transform; direction: ltr;"
    >
      <div
        style="height: 24px; width: 96px;"
      >
        <div
          class="css-m69qwo-MuiStack-root"
          style="position: absolute; left: 0px; top: 0px; height: 24px; width: 24px;"
        >
          <div
            aria-label="Tile 0 0"
            class="MuiBox-root css-7wl9ds"
            data-testid="DefaultTile"
            tabindex="0"
          />
        </div>
        <div
          class="css-m69qwo-MuiStack-root"
          style="position: absolute; left: 24px; top: 0px; height: 24px; width: 24px;"
        >
          <div
            aria-label="Tile 1 0"
            class="MuiBox-root css-7wl9ds"
            data-testid="DefaultTile"
            tabindex="0"
          />
        </div>
        <div
          class="css-m69qwo-MuiStack-root"
          style="position: absolute; left: 48px; top: 0px; height: 24px; width: 24px;"
        >
          <div
            aria-label="Tile 2 0"
            class="MuiBox-root css-7wl9ds"
            data-testid="DefaultTile"
            tabindex="0"
          />
        </div>
        <div
          class="css-m69qwo-MuiStack-root"
          style="position: absolute; left: 72px; top: 0px; height: 24px; width: 24px;"
        >
          <div
            aria-label="Tile 3 0"
            class="MuiBox-root css-7wl9ds"
            data-testid="DefaultTile"
            tabindex="0"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
